source/event_stream.c

```diff
@@ -0,0 +1,1118 @@
+    uint8_t *buffer_alias = buffer;
+
+    for (size_t i = 0; i < headers_count; ++i) {
+        struct aws_event_stream_header_value_pair *header;
+
+        aws_array_list_get_at_ptr(headers, (void **) &header, i);
+        *buffer_alias = (uint8_t) header->header_name_len;
+        buffer_alias++;
+        memcpy(buffer_alias, header->header_name, (size_t) header->header_name_len);
+        buffer_alias += header->header_name_len;
+        *buffer_alias = (uint8_t) header->header_value_type;
+        buffer_alias++;
+        switch (header->header_value_type) {
+            case AWS_EVENT_STREAM_HEADER_BOOL_FALSE:
+            case AWS_EVENT_STREAM_HEADER_BOOL_TRUE:
+                break;
+            case AWS_EVENT_STREAM_HEADER_BYTE:
+                *buffer_alias = header->header_value.static_val[0];
+                buffer_alias++;
+                break;
+                /* additions of integers here assume the endianness conversion has already happened */
+            case AWS_EVENT_STREAM_HEADER_INT16:
+                memcpy(buffer_alias, header->header_value.static_val, sizeof(uint16_t));
+                buffer_alias += sizeof(uint16_t);
+                break;
+            case AWS_EVENT_STREAM_HEADER_INT32:
+                memcpy(buffer_alias, header->header_value.static_val, sizeof(uint32_t));
+                buffer_alias += sizeof(uint32_t);
+                break;
+            case AWS_EVENT_STREAM_HEADER_INT64:
+            case AWS_EVENT_STREAM_HEADER_TIMESTAMP:
+                memcpy(buffer_alias, header->header_value.static_val, sizeof(uint64_t));
+                buffer_alias += sizeof(uint64_t);
+                break;
+            case AWS_EVENT_STREAM_HEADER_BYTE_BUF:
+            case AWS_EVENT_STREAM_HEADER_STRING:
+                aws_write_u16(buffer_alias, header->header_value_len);
+                buffer_alias += sizeof(uint16_t);
+                memcpy(buffer_alias, header->header_value.variable_len_val, header->header_value_len);
+                buffer_alias += header->header_value_len;
+                break;
+            case AWS_EVENT_STREAM_HEADER_UUID:
+                memcpy(buffer_alias, header->header_value.static_val, 16);
+                buffer_alias += header->header_value_len;
+                break;
+        }
+    }
+
+    return buffer_alias - buffer;
+}
+
+/* Get the headers from the buffer, store them in the headers list.
+ * the user's reponsibility to cleanup the list when they are finished with it.
+ * no buffer copies happen here, the lifetime of the buffer, must outlive the usage of the headers.
+ * returns error codes defined in the public interface. */
+int get_headers_from_buffer(struct aws_array_list *headers,
+                            const uint8_t *buffer, size_t headers_len) {
+
+    /* iterate the buffer per header. */
+    const uint8_t *buffer_start = buffer;
+    while ((size_t) (buffer - buffer_start) < headers_len) {
```

Comment:
I don't think this check is sufficient.
`(buffer - buffer_start)` could potentially be less than `headers_len` by 2 and later we try to advance the buffer by more than that.